File: Modules/SMES/Barcodes/SIE.Web.Barcodes/WipBatchs/Commands/ReprintBatchCommand.cs
using SIE.Barcodes.Printables;
using SIE.Common.Prints;
using SIE.Domain;
using SIE.Domain.Validation;
using SIE.Web.Command;
using System;
using System.Collections.Generic;
using SIE.Web.Common.Prints;
using SIE.Barcodes.WipBatchs;
using SIE.Barcodes.Barcodes.ViewModels;

namespace SIE.Web.Barcodes.WipBatchs.Commands
{
    /// <summary>
    /// 批次信息补打
    /// </summary>
    public class ReprintBatchCommand : ListViewCommand
    {
        /// <summary>
        /// 补打命令执行方法
        /// </summary>
        /// <param name="args">args</param>
        /// <param name="scope">scope</param>
        /// <returns>补打结果</returns>
        protected override object Excute(ViewArgs args, string scope)
        {
            var rstPrint = new RstBarcodePrint();
            rstPrint.ErrMsg = string.Empty;

            try
            {
                var entity = args.Data.ToJsonObject<BatchReprintDataModel>();
                var template = RF.GetById<PrintTemplate>(entity.TemplateId);
                Check.NotNull(template, nameof(PrintTemplate));

                if (template.EntityType != typeof(WipBatchPrintable).GetQualifiedName())
                {
                    throw new ValidationException("打印模板的类型必须是生产批次".L10N());
                }

                rstPrint.Type = template.Type;

                var wipBatchs = RT.Service.Resolve<WipBatchController>().ReprintWipBatch(entity.WipBatchIds);
                var report = ReportFactory.Current.GetReportByExtension(template.Type);
                
                var printable = new WipBatchPrintable();

                rstPrint.Url = report.PrintProcess(printable, template.Id, template.Content, () =>
                {
                    return wipBatchs;
                });
            }
            catch (Exception exc)
            {
                rstPrint.ErrMsg = exc.Message;
            }

            return rstPrint;
        }
    }

    /// <summary>
    /// 批次补打数据类
    /// </summary>
    public class BatchReprintDataModel
    {
        /// <summary>
        /// 选中条码Id列表
        /// </summary>
        public List<double> WipBatchIds { get; set; }

        /// <summary>
        /// 补打模板Id
        /// </summary>
        public double TemplateId { get; set; }
    }
}

File: Modules/SMES/Barcodes/SIE.Web.Barcodes/WipBatchs/Commands/ReprintBatchCommand.js
SIE.defineCommand('SIE.Web.Barcodes.WipBatchs.Commands.ReprintBatchCommand', {
    meta: { text: "补打", group: "edit", iconCls: "icon-PrintData icon-blue" },
    canExecute: function (view) {
        var selectModels = view.getSelection();
        if (selectModels.length == 0) return false;
        return true;
    },
    execute: function (view, source) {
        var wipBatchId = view.getCurrent().data.Id;
        SIE.AutoUI.getMeta({
            model: "SIE.Web.Barcodes.WipBatchs.ViewModels.BatchReprintViewModel",
            ignoreCommands: false,
            isDetail: true,
            ignoreQuery: true,
            viewGroup: "DetailsView",
            callback: function (res) {
                var mainBolck;
                if (res.mainBolck)
                    mainBolck = res.mainBolck;
                else
                    mainBolck = res;
                var detailView = SIE.AutoUI.createDetailView(mainBolck);
                detailView.token = view.getToken();
                var entity = new detailView._model();
                entity.setIsSubBatch(false);
                var ui = detailView.getControl();
                SIE.invokeDataQuery({
                    type: "SIE.Web.Barcodes.WipBatchs.WipBatchsDataQueryer",
                    method: "GetReprintInfo",
                    params: [wipBatchId],
                    async: false,
                    token: view.token,
                    callback: function (res) {
                        if (res.Success) {
                            var defaultInfo = res.Result;
                            entity.setTemplateId(defaultInfo.data.items[0].data.TemplateId);
                            entity.setTemplateId_Display(defaultInfo.data.items[0].data.TemplateId_Display);
                            detailView.setData(entity);
                            var win = SIE.Window.show({
                                title: "生产批次补打".t(),
                                width: 480,
                                height: 250,
                                items: ui,
                                id: "ReprintBatch001",
                                callback: function (btn) {
                                    if (btn == "确定".t()) {
                                        var reprintInfo = detailView.getData().data;
                                        if (!reprintInfo.TemplateId || reprintInfo.TemplateId <= 0) {
                                            SIE.Msg.showMessage("打印模板不能为空".L10N());
                                            return false;
                                        }
                                        view.execute({
                                            data: { WipBatchIds: view.getSelectionIds(), TemplateId: reprintInfo.TemplateId },
                                            success: function (res) {
                                                win.close();

                                                var rstPrint = res.Result;

                                                if (rstPrint.ErrMsg !== '') {
                                                    SIE.Msg.showError(rstPrint.ErrMsg);
                                                    return false;
                                                } else {
                                                    win.close();

                                                    var printCmpt = new SIE.Web.Common.Prints.Report.WebReportComponents({
                                                        ReportType: rstPrint.Type,
                                                        ReportData: {
                                                            path: rstPrint.Url, content: rstPrint.Url
                                                        }
                                                    });

                                                    var cfg = printCmpt.getExtTarget();

                                                    if (cfg && cfg.printCallback) {
                                                        cfg.printCallback(printCmpt);
                                                    }
                                                    else {
                                                        var param = printCmpt.getPrintParams();
                                                        if (!printCmpt.hasError()) {
                                                            var printUrl = printCmpt.getPrintUrl();
                                                            if (!printCmpt.hasError())
                                                                CRT.Workbench.showPageDialog({ id: 'ReprintBatch_rpt', text: "生产批次补打".t(), method: 'POST', url: printUrl, params: param });
                                                        }
                                                    }

                                                    view.getSelectionModel().deselectAll();

                                                    view._parent.reloadData();
                                                }
                                            },
                                            error: function (res) {
                                                SIE.Msg.showMessage(res.Message);
                                                return false;
                                            }
                                        });
                                    }
                                }
                            });
                        }
                    }
                });
            }
        });
    }
});